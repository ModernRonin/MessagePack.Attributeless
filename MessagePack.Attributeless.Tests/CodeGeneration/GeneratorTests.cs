using ApprovalTests;
using FluentAssertions;
using MessagePack.Attributeless.CodeGeneration;
using NUnit.Framework;

namespace MessagePack.Attributeless.Tests.CodeGeneration;

[TestFixture]
public class GeneratorTests
{
    class BadType
    {
        public TestActionAttribute UnconfiguredType { get; set; }
    }

    [Test]
    public void Does_not_throw_for_valid_configuration()
    {
        var config = MessagePackSerializer.DefaultOptions.Configure()
            .GraphOf<Samples.IAnimal>()
            .Configuration;

        var underTest = new Generator();

        var action = () => underTest.Generate(config);

        action.Should().NotThrow();
    }

    [Test]
    public void Generate_concrete_type_with_enum()
    {
        var config = MessagePackSerializer.DefaultOptions.Configure()
            .GraphOf<Samples.Leg>()
            .Configuration;

        var result = new Generator().Generate(config);

        Approvals.Verify(result);
    }

    [Test]
    public void Generate_type_hierarchy()
    {
        var config = MessagePackSerializer.DefaultOptions.Configure()
            .GraphOf<Samples.IExtremity>()
            .Configuration;

        var result = new Generator().Generate(config);

        Approvals.Verify(result);
    }

    [Test]
    public void Throws_for_invalid_configuration()
    {
        var config = MessagePackSerializer.DefaultOptions.Configure()
            .GraphOf<BadType>()
            .Configuration;

        var underTest = new Generator();

        var action = () => underTest.Generate(config);

        action.Should()
            .Throw<ArgumentException>()
            .WithMessage(
                "configuration is incomplete - the following types are neither mapped nor handled by MessagePack natively nor enums: NUnit.Framework.TestActionAttribute");
    }
}