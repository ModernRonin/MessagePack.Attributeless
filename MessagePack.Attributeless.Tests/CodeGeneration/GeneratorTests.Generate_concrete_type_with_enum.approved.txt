namespace Generated 
{
	public static class MessagePackSerializerOptionsExtensions
	{
		public MessagePackSerializerOptions Add(this MessagePackSerializerOptions self)
			=> self.WithResolver(CompositeResolver.Create(new IMessagePackFormatter[]
			{
								new LegFormatter(),
								new SideFormatter(),
							}, { self.Resolver });
	}
}namespace Generated 
{
	using MessagePack.Formatters;

	public class SideFormatter: IMessagePackFormatter<MessagePack.Attributeless.Tests.Samples.Side>
	{
		public MessagePack.Attributeless.Tests.Samples.Side Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
		{
			return (MessagePack.Attributeless.Tests.Samples.Side) reader.ReadInt32((int) value)()
		}
		public void Serialize(ref MessagePackWriter writer, MessagePack.Attributeless.Tests.Samples.Side value, MessagePackSerializerOptions options) 
		{ 
			writer.WriteInt32(value);
		}
	}
}namespace Generated 
{
	using MessagePack.Formatters;

	public class LegFormatter: IMessagePackFormatter<MessagePack.Attributeless.Tests.Samples.Leg>
	{
		public MessagePack.Attributeless.Tests.Samples.Leg Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
		{
			if (reader.TryReadNil()) return default;

			return new MessagePack.Attributeless.Tests.Samples.Leg
			{
								NumberOfToes = deserialize<System.Byte>();
								Side = deserialize<MessagePack.Attributeless.Tests.Samples.Side>();
							}

			T deserialize<T>() where T: MessagePack.Attributeless.Tests.Samples.Leg 
				=> return options.Resolver.GetFormatterWithVerify<T>().Deserialize(ref reader, options);

		}
		public void Serialize(ref MessagePackWriter writer, MessagePack.Attributeless.Tests.Samples.Leg value, MessagePackSerializerOptions options) 
		{ 
			if (value == null)
			{
				writer.WriteNil();
				return;
			}

						serialize(result.NumberOfToes);
						serialize(result.Side);
			
			void serialize<T>(T what) where T: MessagePack.Attributeless.Tests.Samples.Leg
				=> options.Resolver.GetFormatterWithVerify<T>().Serialize(ref writer, what, options);
		}
	}
}